File: app/src/test/java/com/jameskbride/eventbusmvpdemo/network/NetworkApiWrapperTest.java
package com.jameskbride.eventbusmvpdemo.network;

import com.jameskbride.eventbusmvpdemo.bus.NetworkErrorEvent;
import com.jameskbride.eventbusmvpdemo.bus.NetworkRequestEvent;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

import java.io.IOException;

import io.reactivex.functions.Consumer;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.verify;
import static org.mockito.MockitoAnnotations.initMocks;

public class NetworkApiWrapperTest {

    @Mock private TameConsumer tameConsumer;

    private NetworkApiWrapper subject;

    private EventBus eventBus;
    private NetworkRequestEvent networkRequestEvent;
    private NetworkErrorEvent networkErrorEvent;

    @Before
    public void setUp() {
        initMocks(this);
        eventBus = EventBus.getDefault();
        networkRequestEvent = new NetworkRequestEvent();

        subject = new NetworkApiWrapper(eventBus, tameConsumer, networkRequestEvent);

        eventBus.register(this);
    }

    @After
    public void tearDown() {
        eventBus.unregister(this);
    }

    @Test
    public void givenAnIOExceptionOccurredThenANetworkErrorEventIsPosted() throws Exception {
        subject.accept(new IOException("Broken pipe"));

        assertEquals(networkRequestEvent, networkErrorEvent.getNetworkRequestEvent());
    }

    @Test
    public void givenANonIOExceptionOccurredThenItDelegatesToTheNextConsumer() throws Exception {
        IllegalAccessException exception = new IllegalAccessException("Some other exception");
        subject.accept(exception);

        verify(tameConsumer).accept(exception);
    }

    @Subscribe
    public void onNetworkErrorEvent(NetworkErrorEvent networkErrorEvent) {
        this.networkErrorEvent = networkErrorEvent;
    }

    private static class TameConsumer implements Consumer<Throwable> {

        @Override
        public void accept(Throwable throwable) throws Exception {

        }
    }
}

File: app/src/test/java/com/jameskbride/eventbusmvpdemo/network/NetworkApiWrapperTest.kt
package com.jameskbride.eventbusmvpdemo.network

import com.jameskbride.eventbusmvpdemo.bus.GetProfileErrorEvent
import com.jameskbride.eventbusmvpdemo.bus.NetworkErrorEvent
import com.jameskbride.eventbusmvpdemo.bus.NetworkRequestEvent
import io.reactivex.functions.Consumer
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.mockito.MockitoAnnotations.initMocks
import java.io.IOException


class NetworkApiWrapperTest {

    private lateinit var subject: NetworkApiWrapper<Throwable>

    private lateinit var eventBus: EventBus
    private lateinit var networkRequestEvent: NetworkRequestEvent
    private lateinit var networkErrorEvent: NetworkErrorEvent

    private var getProfileErrorEventFired: Boolean = false

    @Before
    fun setUp() {
        initMocks(this)
        eventBus = EventBus.getDefault()
        networkRequestEvent = NetworkRequestEvent()

        val errorConsumer = Consumer<Throwable> { error -> eventBus.post(GetProfileErrorEvent()) }

        subject = NetworkApiWrapper(eventBus, errorConsumer, networkRequestEvent)

        eventBus.register(this)
    }

    @After
    fun tearDown() {
        eventBus.unregister(this)
    }

    @Test
    fun givenAnIOExceptionOccurredThenANetworkErrorEventIsPosted() {
        subject.accept(IOException("Broken pipe"))

        assertEquals(networkRequestEvent, networkErrorEvent!!.networkRequestEvent)
    }

    @Test
    fun givenANonIOExceptionOccurredThenItDelegatesToTheNextConsumer() {
        val exception = IllegalAccessException("Some other exception")
        subject.accept(exception)

        assertTrue(getProfileErrorEventFired)
    }

    @Subscribe
    fun onGetProfileErrorEvent(getProfileErrorEvent: GetProfileErrorEvent) {
        this.getProfileErrorEventFired = true
    }

    @Subscribe
    fun onNetworkErrorEvent(networkErrorEvent: NetworkErrorEvent) {
        this.networkErrorEvent = networkErrorEvent
    }
}